frontend/test/pytest/test_measurement_primitives.py

```diff
@@ -25,19 +26,172 @@
     """Test that the sample primitive can be captured into jaxpr."""
 
     def f():
         obs = compbasis_p.bind()
-        return sample_p.bind(obs, shots=5, shape=(5, 0))
+        return sample_p.bind(obs, 5, shape=(5, 0))
 
     jaxpr = jax.make_jaxpr(f)().jaxpr
-    assert jaxpr.eqns[1].primitive == sample_p
-    assert jaxpr.eqns[1].params == {"shape": (5, 0), "shots": 5}
-    assert jaxpr.eqns[1].outvars[0].aval.shape == (5, 0)
+    mlir = lower_jaxpr_to_mlir(jax.make_jaxpr(f)(), "foo")[0]
+    assert (
+        jaxpr
+        == """
+{ lambda ; . let
+    a:AbstractObs(num_qubits=0,primitive=compbasis) = compbasis
+    b:f64[5,0] = sample[shape=(5, 0)] a 5
+  in (b,) }
+"""
+    )
+
+    assert (
+        mlir
+        == """
+module @foo {
+  func.func public @jit_foo() -> tensor<5x0xf64> {
+    %0 = "quantum.compbasis"() : () -> !quantum.obs
+    %c = stablehlo.constant dense<5> : tensor<i64>
+    %1 = "tensor.extract"(%c) : (tensor<i64>) -> i64
+    %2 = "quantum.sample"(%0, %1) : (!quantum.obs, i64) -> tensor<5x0xf64>
+    return %2 : tensor<5x0xf64>
+  }
+}
+"""
+    )
+
+    # assert jaxpr.eqns[1].primitive == sample_p
+    # assert jaxpr.eqns[1].params == {"shape": (5, 0), "shots": 5}
+    # assert jaxpr.eqns[1].outvars[0].aval.shape == (5, 0)
+
+@pytest.mark.xfail(reason="[WIP] Convert to lit test")
+def test_sample_dynamic_shape():
+    """Test that the sample primitive with dynamic shape can be captured into jaxpr."""
+
+    def f(shots):
+        obs = compbasis_p.bind()
+        x = shots + 1
+        # Note that in `primitive.bind(args, kwargs)`, args are treated as jaxpr primitive's
+        # proper arguments, and kwargs are treated as primitive's `params`
+        # Proper primitive arguments are propagated as jaxpr variables,
+        # whereas primitive params are tracers.
+        return sample_p.bind(obs, x, shape=(x, 0))
+
+    jaxpr = jax.make_jaxpr(f)(5).jaxpr
+    mlir = lower_jaxpr_to_mlir(jax.make_jaxpr(f)(5), "foo")[0]
+    assert (
+        jaxpr
+        == """
+{ lambda ; a:i64[]. let
+    b:AbstractObs(num_qubits=0,primitive=compbasis) = compbasis
+    c:i64[] = add a 1
+    d:f64[c,0] = sample[
+      shape=(Traced<ShapedArray(int64[], weak_type=True)>with<DynamicJaxprTrace(level=1/0)>, 0)
+    ] b c
+  in (c, d) }
+"""
+    )
+
+    assert (
+        mlir
+        == """
+module @foo {
+  func.func public @jit_foo(%arg0: tensor<i64>) -> (tensor<i64>, tensor<?x0xf64>) {
+    %0 = "quantum.compbasis"() : () -> !quantum.obs
+    %c = stablehlo.constant dense<1> : tensor<i64>
+    %1 = stablehlo.add %arg0, %c : tensor<i64>
+    %2 = "quantum.sample"(%0, %1) : (!quantum.obs, tensor<i64>) -> tensor<?x0xf64>
+    return %1, %2 : tensor<i64>, tensor<?x0xf64>
+  }
+}
+"""
+    )
+
+    # assert jaxpr.eqns[1].primitive == sample_p
+    # assert jaxpr.eqns[1].params == {"shape": (5, 0), "shots": 5}
+    # assert jaxpr.eqns[1].outvars[0].aval.shape == (5, 0)
+
+
+def test_new_sampleop_still_good_with_backend():
```

Comment:
I would love a test to check that the new IR with dynamic sample result shapes can be accepted by the backend, although outside of calling `replace_ir` and supply a hardcoded IR I can't think of a good alternative.